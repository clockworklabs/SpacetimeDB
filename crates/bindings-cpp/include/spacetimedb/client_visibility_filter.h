#pragma once

#include <string>

namespace SpacetimeDb {

/// A row-level security filter,
/// which can be registered using the SPACETIMEDB_CLIENT_VISIBILITY_FILTER macro.
///
/// Currently, the only valid value for a filter is a Filter::Sql.
/// This is a filter written as a SQL query. Rows that match this query will be made visible to clients.
///
/// The query must be of the form `SELECT * FROM table` or `SELECT table.* from table`,
/// followed by any number of `JOIN` clauses and a `WHERE` clause.
/// If the query includes any `JOIN`s, it must be in the form `SELECT table.* FROM table`.
/// In any case, the query must select all of the columns from a single table, and nothing else.
///
/// SQL queries are not checked for syntactic or semantic validity
/// until they are processed by the SpacetimeDB host.
/// This means that errors in queries used as SPACETIMEDB_CLIENT_VISIBILITY_FILTER rules
/// will be reported during `spacetime publish`, not at compile time.
class Filter {
private:
    const char* sql_text_;

public:
    /// Create a SQL-based client visibility filter
    static Filter Sql(const char* sql) {
        return Filter(sql);
    }

    /// Get the SQL text for this filter
    const char* sql_text() const {
        return sql_text_;
    }

private:
    explicit Filter(const char* sql) : sql_text_(sql) {}
};

} // namespace SpacetimeDb