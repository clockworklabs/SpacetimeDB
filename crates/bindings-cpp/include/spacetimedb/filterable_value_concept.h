#ifndef SPACETIMEDB_FILTERABLE_VALUE_CONCEPT_H
#define SPACETIMEDB_FILTERABLE_VALUE_CONCEPT_H

#include <type_traits>
#include <concepts>
#include <string>
#include "bsatn/types.h"

// Note: This file should be included within the SpacetimeDb namespace context

/**
 * C++20 concept defining types that can be used as index keys.
 * 
 * A type is filterable if it's one of:
 * - Integral types (including bool)
 * - std::string
 * - Identity, ConnectionId, Timestamp
 * - Simple enums (C-style enums without payloads)
 */
template<typename T>
concept FilterableValue = 
    std::integral<T> ||                    // All integer types + bool
    std::same_as<T, std::string> ||        // String
    std::same_as<T, Identity> ||           // Identity  
    std::same_as<T, ConnectionId> ||       // ConnectionId
    std::same_as<T, Timestamp> ||          // Timestamp
    std::same_as<T, I128> ||               // I128
    std::same_as<T, U128> ||               // U128
    std::is_enum_v<T>;                     // Simple enums

/**
 * Helper to validate constraint at compile time
 */
template<typename TableType, FilterableValue auto member_ptr>
struct ValidateIndexConstraint {
    static constexpr bool value = true;
};

/**
 * Compile-time constraint validation helper
 * Use this to validate that a field can have index constraints
 */
template<typename TableType>
struct ConstraintValidator {
    template<auto member_ptr>
    static constexpr void validate_unique() {
        using FieldType = typename std::remove_reference<decltype(((TableType*)nullptr)->*member_ptr)>::type;
        static_assert(FilterableValue<FieldType>,
            "Field cannot have Unique constraint. "
            "Only integers, bool, std::string, Identity, ConnectionId, Timestamp, "
            "and simple enums can have Unique constraints.");
    }
    
    template<auto member_ptr>
    static constexpr void validate_index() {
        using FieldType = typename std::remove_reference<decltype(((TableType*)nullptr)->*member_ptr)>::type;
        static_assert(FilterableValue<FieldType>,
            "Field cannot have Index constraint. "
            "Only integers, bool, std::string, Identity, ConnectionId, Timestamp, "
            "and simple enums can be indexed.");
    }
    
    template<auto member_ptr>
    static constexpr void validate_primary_key() {
        using FieldType = typename std::remove_reference<decltype(((TableType*)nullptr)->*member_ptr)>::type;
        static_assert(FilterableValue<FieldType>,
            "Field cannot be a PrimaryKey. "
            "Only integers, bool, std::string, Identity, ConnectionId, Timestamp, "
            "and simple enums can be primary keys.");
    }
};

#endif // SPACETIMEDB_FILTERABLE_VALUE_CONCEPT_H