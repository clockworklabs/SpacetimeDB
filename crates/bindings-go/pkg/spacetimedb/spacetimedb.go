// Package spacetimedb provides Go bindings for SpacetimeDB.
//
// This package includes core types, reducer framework, schema management,
// BSATN serialization, and utilities for building SpacetimeDB modules in Go.
//
// Core Types:
//   - Identity: Player identity
//   - Timestamp: Point in time
//   - TimeDuration: Duration of time
//   - ScheduleAt: Scheduling specification
//
// Reducer Framework:
//   - ReducerFunction: Interface for reducers
//   - LifecycleFunction: Interface for lifecycle functions
//   - ReducerRegistry: Registry for managing reducers
//   - GenericReducer: Simple reducer implementation
//   - GenericLifecycleFunction: Simple lifecycle function implementation
//
// Schema Management:
//   - TableInfo: Table definition structure
//   - Column: Column definition structure
//   - Index: Index definition structure
//   - TableRegistry: Registry for table definitions
//
// BSATN Serialization:
//   - High-performance binary serialization format
//   - Support for all primitive types (u8, u16, u32, u64, i8, i16, i32, i64, f32, f64, bool, string, bytes)
//   - Advanced collections (arrays, maps, optionals)
//   - Generic encoding/decoding functions
//   - Size calculation utilities
//
// Example usage:
//
//	import "github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb"
//
//	// Use core types
//	identity := spacetimedb.NewIdentity([16]byte{...})
//	timestamp := spacetimedb.NewTimestampFromTime(time.Now())
//
//	// Create table schema
//	table := spacetimedb.NewTableInfo("users")
//	table.Columns = []spacetimedb.TableColumn{
//		spacetimedb.NewPrimaryKeyColumn("id", spacetimedb.TypeU32),
//		spacetimedb.NewColumn("name", spacetimedb.TypeString),
//	}
//	spacetimedb.GlobalRegisterTable(table)
//
//	// Create and register a reducer
//	myReducer := spacetimedb.NewGenericReducer("my_reducer", "Does something", func(ctx *spacetimedb.ReducerContext, args []byte) spacetimedb.ReducerResult {
//		// Your reducer logic here
//		return spacetimedb.NewSuccessResult()
//	})
//	spacetimedb.RegisterReducer(myReducer)
package spacetimedb

// Re-export all components for convenience
import (
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/bsatn"
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/reducers"
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/schema"
	"github.com/clockworklabs/SpacetimeDB/crates/bindings-go/pkg/spacetimedb/types"
)

// Core SpacetimeDB types
type (
	// Identity represents a unique player identity
	Identity = types.Identity

	// Timestamp represents a point in time
	Timestamp = types.Timestamp

	// TimeDuration represents a duration of time
	TimeDuration = types.TimeDuration

	// ScheduleAt represents when a scheduled reducer should be executed
	ScheduleAt = types.ScheduleAt
)

// Reducer framework types
type (
	// LifecycleType represents the type of lifecycle event
	LifecycleType = reducers.LifecycleType

	// ReducerContext contains information about the current reducer execution
	ReducerContext = reducers.ReducerContext

	// ReducerResult represents the result of a reducer execution
	ReducerResult = reducers.ReducerResult

	// ReducerFunction represents a function that can be called as a reducer
	ReducerFunction = reducers.ReducerFunction

	// LifecycleFunction represents a function that can be called for lifecycle events
	LifecycleFunction = reducers.LifecycleFunction

	// GenericReducer provides a simple implementation of ReducerFunction
	GenericReducer = reducers.GenericReducer

	// GenericLifecycleFunction provides a simple implementation of LifecycleFunction
	GenericLifecycleFunction = reducers.GenericLifecycleFunction

	// ReducerRegistry manages all registered reducers and lifecycle functions
	ReducerRegistry = reducers.ReducerRegistry

	// ReducerMetrics tracks performance metrics for reducers
	ReducerMetrics = reducers.ReducerMetrics
)

// Schema framework types
type (
	// TableInfo contains metadata about a SpacetimeDB table
	TableInfo = schema.TableInfo

	// TableColumn represents a table column definition
	TableColumn = schema.Column

	// TableIndex represents a table index definition
	TableIndex = schema.Index

	// TableRegistry manages a collection of table definitions
	TableRegistry = schema.TableRegistry

	// SchemaTableID represents a unique identifier for a SpacetimeDB table in schema
	SchemaTableID = schema.TableID

	// ColumnID represents a unique identifier for a table column
	ColumnID = schema.ColumnID

	// SchemaIndexID represents a unique identifier for a table index in schema
	SchemaIndexID = schema.IndexID

	// IndexType represents the type of index
	IndexType = schema.IndexType

	// TableRegistrationOptions configures table registration behavior
	TableRegistrationOptions = schema.RegistrationOptions

	// TableRegistryStats contains statistics about a table registry
	TableRegistryStats = schema.RegistryStats
)

// Lifecycle constants
const (
	LifecycleInit       = reducers.LifecycleInit
	LifecycleUpdate     = reducers.LifecycleUpdate
	LifecycleConnect    = reducers.LifecycleConnect
	LifecycleDisconnect = reducers.LifecycleDisconnect
)

// Schema constants - SpacetimeDB type names
const (
	TypeU8           = schema.TypeU8
	TypeU16          = schema.TypeU16
	TypeU32          = schema.TypeU32
	TypeU64          = schema.TypeU64
	TypeU128         = schema.TypeU128
	TypeU256         = schema.TypeU256
	TypeI8           = schema.TypeI8
	TypeI16          = schema.TypeI16
	TypeI32          = schema.TypeI32
	TypeI64          = schema.TypeI64
	TypeI128         = schema.TypeI128
	TypeI256         = schema.TypeI256
	TypeF32          = schema.TypeF32
	TypeF64          = schema.TypeF64
	TypeBool         = schema.TypeBool
	TypeString       = schema.TypeString
	TypeBytes        = schema.TypeBytes
	TypeIdentity     = schema.TypeIdentity
	TypeTimestamp    = schema.TypeTimestamp
	TypeTimeDuration = schema.TypeTimeDuration
	TypeScheduleAt   = schema.TypeScheduleAt
)

// Index type constants
const (
	IndexTypeBTree  = schema.IndexTypeBTree
	IndexTypeHash   = schema.IndexTypeHash
	IndexTypeDirect = schema.IndexTypeDirect
)

// Core type constructors
var (
	// Identity constructors
	NewIdentity = types.NewIdentity

	// Timestamp constructors
	NewTimestamp         = types.NewTimestamp
	NewTimestampFromTime = types.NewTimestampFromTime

	// TimeDuration constructors
	NewTimeDuration             = types.NewTimeDuration
	NewTimeDurationFromDuration = types.NewTimeDurationFromDuration

	// ScheduleAt constructors
	NewScheduleAtTime     = types.NewScheduleAtTime
	NewScheduleAtInterval = types.NewScheduleAtInterval
)

// Reducer result constructors
var (
	NewSuccessResult            = reducers.NewSuccessResult
	NewSuccessResultWithMessage = reducers.NewSuccessResultWithMessage
	NewErrorResult              = reducers.NewErrorResult
	NewErrorResultWithMessage   = reducers.NewErrorResultWithMessage
)

// Reducer and lifecycle constructors
var (
	NewGenericReducer           = reducers.NewGenericReducer
	NewGenericLifecycleFunction = reducers.NewGenericLifecycleFunction
	NewReducerRegistry          = reducers.NewReducerRegistry
	NewReducerMetrics           = reducers.NewReducerMetrics
)

// Schema constructors
var (
	// Table constructors
	NewTableInfo     = schema.NewTableInfo
	NewTableRegistry = schema.NewTableRegistry

	// Column constructors
	NewColumn           = schema.NewColumn
	NewPrimaryKeyColumn = schema.NewPrimaryKeyColumn
	NewAutoIncColumn    = schema.NewAutoIncColumn

	// Index constructors
	NewIndex       = schema.NewIndex
	NewBTreeIndex  = schema.NewBTreeIndex
	NewUniqueIndex = schema.NewUniqueIndex

	// Options constructor
	DefaultRegistrationOptions = schema.DefaultRegistrationOptions
)

// Global registry functions - Reducers
var (
	RegisterReducer           = reducers.RegisterReducer
	RegisterLifecycleFunction = reducers.RegisterLifecycleFunction
	GetReducer                = reducers.GetReducer
	GetLifecycleFunction      = reducers.GetLifecycleFunction
)

// Global registry functions - Tables
var (
	GlobalRegisterTable     = schema.GlobalRegister
	GlobalRegisterAllTables = schema.GlobalRegisterAll
	GlobalGetTable          = schema.GlobalGetTable
	GlobalGetTableByID      = schema.GlobalGetTableByID
	GlobalMustGetTable      = schema.GlobalMustGetTable
	GlobalHasTable          = schema.GlobalHasTable
	GlobalGetAllTables      = schema.GlobalGetAllTables
	GlobalGetTableNames     = schema.GlobalGetTableNames
	GlobalTableCount        = schema.GlobalCount
	GlobalClearTables       = schema.GlobalClear
	GlobalValidateAllTables = schema.GlobalValidateAll
	GlobalGetTableStats     = schema.GlobalGetStats
	GetGlobalTableRegistry  = schema.GetGlobalRegistry
)

// BSATN serialization functions
var (
	// Primitive type encoders
	BsatnEncodeU8     = bsatn.EncodeU8
	BsatnEncodeU16    = bsatn.EncodeU16
	BsatnEncodeU32    = bsatn.EncodeU32
	BsatnEncodeU64    = bsatn.EncodeU64
	BsatnEncodeI8     = bsatn.EncodeI8
	BsatnEncodeI16    = bsatn.EncodeI16
	BsatnEncodeI32    = bsatn.EncodeI32
	BsatnEncodeI64    = bsatn.EncodeI64
	BsatnEncodeF32    = bsatn.EncodeF32
	BsatnEncodeF64    = bsatn.EncodeF64
	BsatnEncodeBool   = bsatn.EncodeBool
	BsatnEncodeString = bsatn.EncodeString
	BsatnEncodeBytes  = bsatn.EncodeBytes

	// Primitive type decoders
	BsatnDecodeU8     = bsatn.DecodeU8
	BsatnDecodeU16    = bsatn.DecodeU16
	BsatnDecodeU32    = bsatn.DecodeU32
	BsatnDecodeU64    = bsatn.DecodeU64
	BsatnDecodeI8     = bsatn.DecodeI8
	BsatnDecodeI16    = bsatn.DecodeI16
	BsatnDecodeI32    = bsatn.DecodeI32
	BsatnDecodeI64    = bsatn.DecodeI64
	BsatnDecodeF32    = bsatn.DecodeF32
	BsatnDecodeF64    = bsatn.DecodeF64
	BsatnDecodeBool   = bsatn.DecodeBool
	BsatnDecodeString = bsatn.DecodeString
	BsatnDecodeBytes  = bsatn.DecodeBytes

	// Array encoders/decoders
	BsatnEncodeU32Array    = bsatn.EncodeU32Array
	BsatnDecodeU32Array    = bsatn.DecodeU32Array
	BsatnEncodeStringArray = bsatn.EncodeStringArray
	BsatnDecodeStringArray = bsatn.DecodeStringArray
	BsatnEncodeF64Array    = bsatn.EncodeF64Array
	BsatnDecodeF64Array    = bsatn.DecodeF64Array

	// Generic collection functions
	BsatnEncodeArray    = bsatn.EncodeArray[any]
	BsatnDecodeArray    = bsatn.DecodeArray[any]
	BsatnEncodeOptional = bsatn.EncodeOptional[any]
	BsatnDecodeOptional = bsatn.DecodeOptional[any]
	BsatnEncodeMap      = bsatn.EncodeMap[string, any]
	BsatnDecodeMap      = bsatn.DecodeMap[string, any]

	// Size calculation functions
	BsatnSizeU8          = bsatn.SizeU8
	BsatnSizeU16         = bsatn.SizeU16
	BsatnSizeU32         = bsatn.SizeU32
	BsatnSizeU64         = bsatn.SizeU64
	BsatnSizeI8          = bsatn.SizeI8
	BsatnSizeI16         = bsatn.SizeI16
	BsatnSizeI32         = bsatn.SizeI32
	BsatnSizeI64         = bsatn.SizeI64
	BsatnSizeF32         = bsatn.SizeF32
	BsatnSizeF64         = bsatn.SizeF64
	BsatnSizeBool        = bsatn.SizeBool
	BsatnSizeString      = bsatn.SizeString
	BsatnSizeBytes       = bsatn.SizeBytes
	BsatnSizeU32Array    = bsatn.SizeU32Array
	BsatnSizeStringArray = bsatn.SizeStringArray
	BsatnSizeF64Array    = bsatn.SizeF64Array

	// Utility functions
	BsatnToBytes              = bsatn.ToBytes
	BsatnFromBytes            = bsatn.FromBytes
	BsatnU32ArrayToBytes      = bsatn.U32ArrayToBytes
	BsatnU32ArrayFromBytes    = bsatn.U32ArrayFromBytes
	BsatnStringArrayToBytes   = bsatn.StringArrayToBytes
	BsatnStringArrayFromBytes = bsatn.StringArrayFromBytes
)

// Package information
const (
	Version = "0.3.0"
	Name    = "SpacetimeDB Go Bindings"
)

// GetVersion returns the current version of the bindings
func GetVersion() string {
	return Version
}

// GetPackageName returns the package name
func GetPackageName() string {
	return Name
}

// GetFeatures returns a list of supported features
func GetFeatures() []string {
	return []string{
		"Core Types (Identity, Timestamp, TimeDuration, ScheduleAt)",
		"Reducer Framework (Registration, Execution, Metrics)",
		"Schema Management (Tables, Columns, Indexes)",
		"BSATN Serialization (Binary, High-Performance)",
		"Collection Support (Arrays, Maps, Optionals)",
		"Global Registries (Thread-safe, Validation)",
		"JSON Serialization",
		"WASM Support",
	}
}
