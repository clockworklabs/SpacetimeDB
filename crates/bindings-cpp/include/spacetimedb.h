#pragma once

/**
 * SpacetimeDB C++20 Module API - Single Include Header
 * 
 * Complete, optimized API for building SpacetimeDB modules with support for:
 * - Tables with field constraints
 * - BSATN serialization
 * - Reducers (user-defined and lifecycle)
 * - Database operations
 * - Logging and debugging
 * 
 * =============================================================================
 * USAGE EXAMPLES
 * =============================================================================
 * 
 * Table Definition:
 *    struct User {
 *        uint32_t id;
 *        std::string username;
 *        std::string email;
 *    };
 *    SPACETIMEDB_STRUCT(User, id, username, email)
 *    SPACETIMEDB_TABLE(User, "users", Public,
 *        PrimaryKeyAutoInc(id),
 *        Unique(email)
 *    )
 * 
 * Lifecycle Reducers:
 *    SPACETIMEDB_INIT(init_module) {
 *        LOG_INFO("Module initialized!");
 *    }
 * 
 * User Reducers:
 *    void add_user(ReducerContext ctx, std::string username, std::string email) {
 *        User user{0, username, email}; // id auto-generated
 *        ctx.db.table<User>("users").insert(user);
 *    }
 *    REGISTER_REDUCER(add_user, add_user)
 */

// =============================================================================
// STANDARD LIBRARY DEPENDENCIES
// =============================================================================

#include <cstdint>
#include <string>
#include <vector>
#include <optional>
#include <array>
#include <string_view>
#include <concepts>
#include <functional>
#include <type_traits>
#include <span>

// =============================================================================
// CORE SPACETIMEDB TYPE SYSTEM
// =============================================================================

// Core types and BSATN support
#include "spacetimedb/bsatn/types.h"
#include "spacetimedb/bsatn/timestamp.h"
#include "spacetimedb/bsatn/time_duration.h"
#include "spacetimedb/bsatn/writer.h"
#include "spacetimedb/bsatn/reader.h"
#include "spacetimedb/bsatn/algebraic_type.h"
#include "spacetimedb/bsatn/traits.h"
#include "spacetimedb/bsatn/sum_type.h"
#include "spacetimedb/bsatn/schedule_at.h"

// Autogenerated types for module definition
#include "spacetimedb/internal/autogen/SumType.g.h"  // Complete SumType definition first
#include "spacetimedb/internal/autogen/RawModuleDefV9.g.h"
#include "spacetimedb/internal/autogen/RawTableDefV9.g.h"
#include "spacetimedb/internal/autogen/Typespace.g.h"
#include "spacetimedb/internal/autogen/ProductType.g.h"
#include "spacetimedb/internal/autogen/ProductTypeElement.g.h"
#include "spacetimedb/internal/autogen/TableAccess.g.h"
#include "spacetimedb/internal/autogen/TableType.g.h"

// =============================================================================
// MODULE SYSTEM INFRASTRUCTURE
// =============================================================================

// Core module infrastructure
#include "spacetimedb/internal/Module.h"
#include "spacetimedb/internal/field_registration.h"

// Logging system
#include "spacetimedb/logger.h"

// =============================================================================
// TABLE AND CONSTRAINT SYSTEM
// =============================================================================

// Type extensions for special types
#include "spacetimedb/bsatn/type_extensions.h"

// Table operations and constraint support
#include "spacetimedb/table.h"
#include "spacetimedb/database.h"
#include "spacetimedb/table_with_constraints.h"

// =============================================================================
// REDUCER SYSTEM
// =============================================================================

// Reducer context and macros
#include "spacetimedb/reducer_context.h"
#include "spacetimedb/reducer_macros.h"
#include "spacetimedb/reducer_error.h"

// Client visibility filters
#include "spacetimedb/client_visibility_filter.h"
#include "spacetimedb/rls.h"

// =============================================================================
// CONVENIENCE ALIASES AND COMPATIBILITY
// =============================================================================

// Convenience aliases for table access levels
namespace SpacetimeDb {
    // Re-export table access constants for convenience
    constexpr auto Public = Internal::TableAccess::Public;
    constexpr auto Private = Internal::TableAccess::Private;
}

// =============================================================================
// UNIFIED MACRO SYSTEM
// =============================================================================

// Use unified macro system
#include "spacetimedb/macros.h"


// =============================================================================
// MODULE DEFINITION UTILITIES
// =============================================================================

// Forward declarations for module system compatibility
namespace spacetimedb {
    // Template specialization support
    template<typename T> struct table_metadata;
    
    namespace internal {
        /**
         * @brief Serialize the current module definition to BSATN
         * @return Serialized module definition
         */
        inline std::vector<uint8_t> serialize_module_def() {
            auto& module_def = SpacetimeDb::Internal::Module::GetModuleDef();
            return module_def.serialize();
        }
    }

    // =============================================================================
    // CONVENIENCE TYPE ALIASES
    // =============================================================================
    
    // Vector type aliases for SpacetimeDB types
    using VecTimeDuration = std::vector<::SpacetimeDb::TimeDuration>;
}

// =============================================================================
// DOCUMENTATION AND USAGE NOTES
// =============================================================================

/**
 * CONSTRAINT TYPES REFERENCE:
 * 
 * Basic Constraints:
 * - PrimaryKey(field) - Primary key constraint (unique, non-null)
 * - Unique(field) - Unique constraint (allows null)
 * - Index(field) - Index for fast lookups
 * - AutoInc(field) - Auto-incrementing field
 * 
 * Combination Constraints:
 * - PrimaryKeyAutoInc(field) - Primary key with auto-increment
 * - UniqueAutoInc(field) - Unique field with auto-increment  
 * - IndexAutoInc(field) - Indexed field with auto-increment
 * 
 * LIFECYCLE REDUCERS:
 * - SPACETIMEDB_INIT(name) - Called when module is initialized
 * - SPACETIMEDB_CLIENT_CONNECTED(name) - Called when client connects
 * - SPACETIMEDB_CLIENT_DISCONNECTED(name) - Called when client disconnects
 * 
 * LOGGING LEVELS:
 * - LOG_DEBUG(msg) - Debug information
 * - LOG_INFO(msg) - General information
 * - LOG_WARN(msg) - Warning messages
 * - LOG_ERROR(msg) - Error messages
 * - LOG_PANIC(msg) - Critical errors (may terminate)
 */

// =============================================================================
// BSATN IMPLEMENTATION INCLUDES
// =============================================================================

// Include BSATN implementation files after all headers are defined
#include "spacetimedb/bsatn/types_impl.h"
#include "spacetimedb/bsatn/schedule_at_impl.h"
#include "spacetimedb/enum_macro.h"