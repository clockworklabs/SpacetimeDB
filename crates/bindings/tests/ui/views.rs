use spacetimedb::{reducer, table, view, AnonymousViewContext, Identity, ReducerContext, ViewContext};

#[table(name = test)]
struct Test {
    #[unique]
    id: u32,
    #[index(btree)]
    x: u32,
}

#[reducer]
fn view_handle_no_iter(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `iter()`
    for _ in read_only.db.test().iter() {}
}

#[reducer]
fn view_handle_no_count(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `count()`
    let _ = read_only.db.test().count();
}

#[reducer]
fn view_handle_no_insert(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `insert()`
    read_only.db.test().insert(Test { id: 0, x: 0 });
}

#[reducer]
fn view_handle_no_try_insert(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `try_insert()`
    read_only.db.test().try_insert(Test { id: 0, x: 0 });
}

#[reducer]
fn view_handle_no_delete(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `delete()`
    read_only.db.test().delete(Test { id: 0, x: 0 });
}

#[reducer]
fn read_only_unqiue_index_no_delete(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: unique read-only index does not expose `delete()`
    read_only.db.test().id().delete(&0);
}

#[reducer]
fn read_only_unqiue_index_no_update(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: unique read-only index does not expose `update()`
    read_only.db.test().id().update(Test { id: 0, x: 0 });
}

#[reducer]
fn read_only_btree_index_no_delete(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: read-only btree index does not expose `delete()`
    read_only.db.test().x().delete(0u32..);
}

#[table(name = player)]
struct Player {
    #[unique]
    identity: Identity,
}

struct NotSpacetimeType {}

/// Private views not allowed; must be `#[view(public, ...)]`
#[view(name = view_def_no_public)]
fn view_def_no_public(_: &ViewContext) -> Vec<Player> {
    vec![]
}

/// Duplicate `public`
#[view(name = view_def_dup_public, public, public)]
fn view_def_dup_public() -> Vec<Player> {
    vec![]
}

/// Duplicate `name`
#[view(name = view_def_dup_name, name = view_def_dup_name, public)]
fn view_def_dup_name() -> Vec<Player> {
    vec![]
}

/// Unsupported attribute arg
#[view(name = view_def_unsupported_arg, public, anonymous)]
fn view_def_unsupported_arg() -> Vec<Player> {
    vec![]
}

/// A `ViewContext` is required
#[view(name = view_def_no_context, public)]
fn view_def_no_context() -> Vec<Player> {
    vec![]
}

/// A `ViewContext` is required
#[view(name = view_def_wrong_context, public)]
fn view_def_wrong_context(_: &ReducerContext) -> Vec<Player> {
    vec![]
}

/// Must pass the `ViewContext` by ref
#[view(name = view_def_pass_context_by_value, public)]
fn view_def_pass_context_by_value(_: ViewContext) -> Vec<Player> {
    vec![]
}

/// The view context must be the first parameter
#[view(name = view_def_wrong_context_position, public)]
fn view_def_wrong_context_position(_: &u32, _: &ViewContext) -> Vec<Player> {
    vec![]
}

/// Must return `Vec<T>` or `Option<T>` where `T` is a SpacetimeType
#[view(name = view_def_no_return, public)]
fn view_def_no_return(_: &ViewContext) {}

/// Must return `Vec<T>` or `Option<T>` where `T` is a SpacetimeType
#[view(name = view_def_wrong_return, public)]
fn view_def_wrong_return(_: &ViewContext) -> Player {
    Player {
        identity: Identity::ZERO,
    }
}

/// Must return `Vec<T>` or `Option<T>` where `T` is a SpacetimeType
#[view(name = view_def_returns_not_a_spacetime_type, public)]
fn view_def_returns_not_a_spacetime_type(_: &AnonymousViewContext) -> Option<NotSpacetimeType> {
    None
}

/// Cannot use a view as a scheduled function
#[spacetimedb::table(name = scheduled_table, scheduled(scheduled_table_view))]
struct ScheduledTable {
    #[primary_key]
    #[auto_inc]
    scheduled_id: u64,
    scheduled_at: spacetimedb::ScheduleAt,
    x: u8,
    y: u8,
}

/// Cannot use a view as a scheduled function
#[view(name = scheduled_table_view, public)]
fn scheduled_table_view(_: &ViewContext, _args: ScheduledTable) -> Vec<Player> {
    vec![]
}

fn main() {}
