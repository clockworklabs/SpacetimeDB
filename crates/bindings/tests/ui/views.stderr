error[E0599]: no method named `iter` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:15:34
   |
15 |     for _ in read_only.db.test().iter() {}
   |                                  ^^^^ method not found in `&test__ViewHandle`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `iter`, perhaps you need to implement one of them:
           candidate #1: `Table`
           candidate #2: `bitflags::traits::Flags`

error[E0599]: `&test__ViewHandle` is not an iterator
  --> tests/ui/views.rs:22:33
   |
3  | #[table(name = test)]
   | ------------------- doesn't satisfy `test__ViewHandle: Iterator`
...
22 |     let _ = read_only.db.test().count();
   |                                 ^^^^^ `&test__ViewHandle` is not an iterator
   |
   = note: the following trait bounds were not satisfied:
           `&test__ViewHandle: Iterator`
           which is required by `&mut &test__ViewHandle: Iterator`
           `test__ViewHandle: Iterator`
           which is required by `&mut test__ViewHandle: Iterator`
note: the trait `Iterator` must be implemented
  --> $RUST/core/src/iter/traits/iterator.rs
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `count`, perhaps you need to implement one of them:
           candidate #1: `Table`
   = note: the trait `Iterator` defines an item `count`, but is explicitly unimplemented

error[E0599]: no method named `insert` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:29:25
   |
29 |     read_only.db.test().insert(Test { id: 0, x: 0 });
   |                         ^^^^^^ method not found in `&test__ViewHandle`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `insert`, perhaps you need to implement one of them:
           candidate #1: `Table`
           candidate #2: `bitflags::traits::Flags`
           candidate #3: `ppv_lite86::types::Vec2`
           candidate #4: `ppv_lite86::types::Vec4`

error[E0599]: no method named `try_insert` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:36:25
   |
36 |     read_only.db.test().try_insert(Test { id: 0, x: 0 });
   |                         ^^^^^^^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `try_insert`, perhaps you need to implement it:
           candidate #1: `Table`
help: there is a method `try_into` with a similar name, but with different arguments
  --> $RUST/core/src/convert/mod.rs

error[E0599]: no method named `delete` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:43:25
   |
43 |     read_only.db.test().delete(Test { id: 0, x: 0 });
   |                         ^^^^^^ method not found in `&test__ViewHandle`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `delete`, perhaps you need to implement it:
           candidate #1: `Table`

error[E0599]: no method named `delete` found for struct `UniqueColumnReadOnly` in the current scope
  --> tests/ui/views.rs:50:30
   |
50 |     read_only.db.test().id().delete(&0);
   |                              ^^^^^^ method not found in `UniqueColumnReadOnly<test__TableHandle, u32, id>`

error[E0599]: no method named `update` found for struct `UniqueColumnReadOnly` in the current scope
  --> tests/ui/views.rs:57:30
   |
57 |     read_only.db.test().id().update(Test { id: 0, x: 0 });
   |                              ^^^^^^ method not found in `UniqueColumnReadOnly<test__TableHandle, u32, id>`

error[E0599]: no method named `delete` found for struct `RangedIndexReadOnly` in the current scope
  --> tests/ui/views.rs:64:29
   |
64 |     read_only.db.test().x().delete(0u32..);
   |                             ^^^^^^ method not found in `RangedIndexReadOnly<test__TableHandle, (u32,), x>`
